src/transformers/models/jukebox/modeling_jukebox.py

```diff
@@ -0,0 +1,3133 @@
+        width,
+        n_ctx,
+        num_heads,
+        attn_dropout=0.0,
+        resid_dropout=0.0,
+        afn="gelu",
+        scale=True,
+        mask=False,
+        zero_out=False,
+        init_scale=1.0,
+        res_scale=1.0,
+        m_attn=0.25,
+        m_mlp=1.0,
+        attn_func=0,
+        blocks=None,
+        spread=None,
+        encoder_dims=None,
+        lyric_enc_len=None,
+    ):
+        super().__init__()
+        self.attn = JukeboxAttention(
+            width=width,
+            n_ctx=n_ctx,
+            hidden_dim=int(m_attn * width),
+            num_heads=num_heads,
+            attn_dropout=attn_dropout,
+            resid_dropout=resid_dropout,
+            scale=scale,
+            mask=mask,
+            zero_out=zero_out,
+            init_scale=init_scale,
+            attn_func=attn_func,
+            blocks=blocks,
+            spread=spread,
+            encoder_dims=encoder_dims,
+            lyric_enc_len=lyric_enc_len,
+        )
+
+        self.layer_norm_0 = JukeboxLayerNorm(width)
+        self.mlp = JukeboxMLP(
+            width=width,
+            hidden_dim=int(m_mlp * width),
+            resid_dropout=resid_dropout,
+            afn=afn,
+            zero_out=zero_out,
+            init_scale=init_scale,
+        )
+        self.layer_norm_1 = JukeboxLayerNorm(width)
+        self.res_scale = res_scale
+
+        self.width = width
+        self.attn_func = attn_func
+
+    def forward(self, hidden_states, lyric_encoder_states, sample=False):
+        residuals = hidden_states
+        hidden_states = self.layer_norm_0(hidden_states)
+        hidden_states = self.attn(hidden_states, lyric_encoder_states, sample)
+
+        output_states = self.layer_norm_1(residuals + hidden_states)
+        output_states = self.mlp(output_states)
+        if self.res_scale == 1.0:
+            output = residuals + hidden_states + output_states
+        else:
+            output = residuals + self.res_scale * (hidden_states + output_states)
+        return output
+
+
+class JukeboxTransformer(nn.Module):
+    def __init__(
+        self,
+        width,
+        n_ctx,
+        num_heads,
+        n_depth,
+        attn_dropout=0.0,
+        resid_dropout=0.0,
+        afn="gelu",
+        scale=True,
+        mask=False,
+        zero_out=False,
+        init_scale=1.0,
+        res_scale=False,
+        m_attn=0.25,
+        m_mlp=1.0,
+        attn_order=0,
+        blocks=None,
+        spread=None,
+        encoder_dims=None,
+        lyric_enc_len=None,
+    ):
+        super().__init__()
+        self.width = width
+        self.n_ctx = n_ctx
+        self.encoder_dims = encoder_dims
+        self.blocks = blocks
+        if blocks is not None:
+            self.block_ctx = n_ctx // blocks
+        self.lyric_enc_len = lyric_enc_len
+        self.num_heads = num_heads
+
+        res_scale = 1.0 / n_depth if res_scale else 1.0
+
+        # Orders of attn_func
+        attn_func = {
```

Comment:
Wow!

Reply:
should we maybe move this out into a function:

```python
attn_func = self.get_attn_func(attn_order)
```
? 
Just so that one doesn't need to worry about understanding what's going on there?